metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+        using kmer_t = typename std::remove_reference<decltype(dict)>::type::kmer_t;
+        sshash::neighbourhood<kmer_t> nb = dict.kmer_backward_neighbours(kmer.c_str(), true);
+        for (size_t i = 0; i < kmer_t::alphabet_size; i++) {
+            if (nb.backward[i].kmer_id != sshash::constants::invalid_uint64) {
+                callback(sshash_to_graph_index(nb.backward[i].kmer_id), kmer_t::alphabet[i],
+                         nb.backward[i].kmer_orientation);
+            }
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        call_impl(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        call_impl(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        call_impl(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+size_t DBGSSHash::outdegree(node_index node) const {
+    size_t res = 0;
+    adjacent_outgoing_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+size_t DBGSSHash::indegree(node_index node) const {
+    size_t res = 0;
+    adjacent_incoming_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+void DBGSSHash::call_nodes(
+        const std::function<void(node_index)>& callback,
+        const std::function<bool()> &terminate) const {
+    for (size_t node_idx = 1; !terminate() && node_idx <= dict_->size(); ++node_idx) {
+        callback(node_idx);
+        if (mode_ == CANONICAL && !terminate()) {
+            size_t rc_node_idx = reverse_complement(node_idx);
+            if (rc_node_idx != node_idx)
+                callback(rc_node_idx);
+        }
+    }
+}
+
+DBGSSHash::node_index DBGSSHash::kmer_to_node(std::string_view kmer) const {
+    if (!num_nodes())
+        return npos;
+
+    auto res = dict_->lookup_advanced(kmer.data(), mode_ == CANONICAL);
+    node_index node = sshash_to_graph_index(res.kmer_id);
+    return res.kmer_orientation ? reverse_complement(node) : node;
+}
+
+std::pair<DBGSSHash::node_index, bool>
+DBGSSHash::kmer_to_node_with_rc(std::string_view kmer) const {
+    if (!num_nodes())
+        return std::make_pair(npos, false);
+
+    auto res = dict_->lookup_advanced(kmer.data(), true);
+    return std::make_pair(sshash_to_graph_index(res.kmer_id), res.kmer_orientation);
+}
+
+std::string DBGSSHash::get_node_sequence(node_index node) const {
+    std::string str_kmer(k_, ' ');
+    node_index node_canonical = node > dict_->size() ? node - dict_->size() : node;
+    uint64_t ssh_idx = graph_index_to_sshash(node_canonical);
+    dict_->access(ssh_idx, str_kmer.data());
+
+    if (node > node_canonical)
+        ::reverse_complement(str_kmer.begin(), str_kmer.end());
+
+    return str_kmer;
+}
+
+void DBGSSHash::serialize(std::ostream& out) const {
+    // TODO
+    throw std::runtime_error("serialize to stream not implemented");
```

Comment:
nit: `std::logic_error`?

Reply:
I'm not sure if `std::logic_error` is appropriate here. The only reason why this is not supported is because the `essentials::saver` doesn't have a constructor that takes an `std::ostream`, though it could.

Reply:
We can maybe open a pull request there.

Reply:
I wrote a custom saver and loader so we can implement these.